### DISCUSSION TITLE
How do I create a function to automatically add values to a table after a user authenticates with nextauth

### DISCUSSION THREAD

**Discussion by hsheikhali1** [supabase/supabase#27114] [Questions]
So I am trying to create a new user in my students table after they authenticate with NextAuth. NextAuth with supabase creates a new user record within the next_auth.users table and I wanted to create a function much like the example given for when a user authenticates with supabase auth, it creates a new record in the profiles table.

I'm trying to understand how I would be able to figure out what sort of data is available after a new user is added to next_auth.users table. Is the data usually the same across the different providers? I plan on have more than one provider available.

Here is my function

```postgres
create function public.handle_new_student()
returns trigger
language plpgsql
security definer set search_path = ''
as $$
begin
  insert into public.students (id, display_name, email)
  values (new.id, new.name, new.email); -- is this correct? how do I know what values are available on `new` when a users is created in `next_auth.users`?
  return new;
end;
$$;
```

and here is my trigger
```postgres
-- trigger the function every time a user is created in next_auth.users
create trigger on_next_auth_user_created
  after insert on next_auth.users
  for each row execute procedure public.handle_new_student();
```

**Comment by GaryAustin1** [comment]
Your idea is correct on using a trigger function.
Seems like your question should be to nextauth group though as to what they provide into their schema and where.

**Reply by hsheikhali1** [reply]
Yeah.. I haven't really seen any docs for this kind of thing.. I essentially want to keep that `next_auth.users` table and my `students` table in sync.

**Reply by GaryAustin1** [reply]
You would need an update trigger then also to keep in your public table in sync (assuming nextauth updates it).

**Reply by hsheikhali1** [reply]
Hmm, when I say `in sync` i really just mean the very first time someone authenticates with next_auth, I need to create a populate my students table as well.. from there I will manage stuff like profile details. I'm trying to mimic exactly what is done for the users table for supabase auth

**Reply by hsheikhali1** [reply]
@GaryAustin1 btw. This is what the next_auth.users table looks like

```postgres
--
-- Create users table
--
CREATE TABLE IF NOT EXISTS next_auth.users
(
    id uuid NOT NULL DEFAULT gen_random_uuid(),
    name text,
    email text,
    "emailVerified" timestamp with time zone,
    image text,
    CONSTRAINT users_pkey PRIMARY KEY (id),
    CONSTRAINT email_unique UNIQUE (email)
);
```

I did some digging and I realised that supabase suggest doing the following for their user management example
```postgres
  insert into public.profiles (id, first_name, last_name)
  values (new.id, new.raw_user_meta_data ->> 'first_name', new.raw_user_meta_data ->> 'last_name');
  return new;
```
where `new.raw_user_meta_data` is a `jsonb` type column. 

Given this is how it works for the auth.users table, shouldn't it be something similar for mine? Instead of `new.raw_user_meta_data ->> 'first_name'` it would be `new.name` ?

**Reply by GaryAustin1** [reply]
Yes, my comment was your idea for the trigger function is correct.  I'm not syntax checking, but it should work similar to the Supabase auth.users trigger.  You would nave new.name and new.email.  Basically all the columns in their table with new. to use them.

**Reply by hsheikhali1** [reply]
Yeah I know, I just wanted to confirm my understanding of how to pull columns from `next_auth.users` was correct.. Not sure if this is a next auth issue or a supabase issue but it seems like the adapter keeps failing.

**Reply by GaryAustin1** [reply]
You have the right idea.  Look in the Postgres logs for errors.  Also you can turn off the trigger to see if the error is in your nextauth side or your trigger function.

**Reply by hsheikhali1** [reply]
One step ahead of you mate! I don't see any specific log errors related to failed upsert attempts but when I turn off the trigger the next auth flow (including creating a new user in `next_auth.users` table) works like a charm. It's something about the trigger function it doesn't like.

**Reply by GaryAustin1** [reply]
If your code is like you showed at start, I don't see anything that stands out as wrong.  You have not shown your student table.
Please double check the dashboard Postgres logs for any error around the time of the operation.  Also if you show your student table and confirm your code is the same as you originally posted I'll glance at, but I'm about out of time right now.

**Reply by hsheikhali1** [reply]
I'm using Prisma so here is what my schema looks like for students.

```
model student {
  id           String    @id @default(dbgenerated("gen_random_uuid()")) @db.Uuid
  display_name String
  email        String
  createdAt    DateTime  @default(now())
  updatedAt    DateTime? @updatedAt
}
```

Here is the sql definition in supabase

```postgres
create table
  public.student (
    id uuid not null default gen_random_uuid (),
    display_name text not null,
    email text not null,
    "createdAt" timestamp without time zone not null default current_timestamp,
    "updatedAt" timestamp without time zone null,
    constraint student_pkey primary key (id)
  ) tablespace pg_default;
```

**Reply by GaryAustin1** [reply]
Agrees with what you have in the function.  

You'll really need to check the logs closely.  

Prisma adds another wrinkle to the whole thing.  You should always disclose you are using that when asking for help here or Discord.  Prisma can change the grants on the public schema.  I'm not saying that is the case, or impacting here, as you have not found the error.  You'll have to find that or be exactly clear on what is going wrong.  See troubleshooting here:  https://supabase.com/partners/integrations/prisma.

Also you can use `raise log 'function new = %',new;` between Begin and End and look in the Postgres logs to see if new is what you expect.

**Reply by hsheikhali1** [reply]
@GaryAustin1 Yeah I should've mentioned that. I will try your suggestion and see if what I see is what I expect!
